pex/resolver.py

```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
   return Platform.create(platform) if platform and platform != 'current' else None
 
 
+class AtomicDirectory(namedtuple('AtomicDirectory', ['work_dir', 'target_dir'])):
+  @classmethod
+  def for_target_dir(cls, target_dir):
+    return cls(work_dir='{}.{}'.format(target_dir, uuid4().hex), target_dir=target_dir)
+
+  @property
+  def is_finalized(self):
+    return os.path.exists(self.target_dir)
+
+  def finalize(self):
+    if self.is_finalized:
```

Comment:
For the `finalize()` method here, it seems valid to perform this check without explicit inter-process locking, but I'm concerned that propagating the result of `.is_finalized` elsewhere may make things harder to reason about (unless we're absolutely sure that the resulting directory will never be moved or deleted). In particular, for this segment later on:
```python
      if not build_result.is_built:
        TRACER.log('Building {} to {}'.format(build_request.source_path, build_result.dist_dir))
        unsatisfied_build_requests.append(build_request)
      else:
        TRACER.log('Using cached build of {} at {}'
                   .format(build_request.source_path, build_result.dist_dir))
        install_requests.extend(build_result.finalize_build())
```
I think this code seems mostly legitimate, but what's a little confusing to me is how `AtomicDirectory#finalize()` will no-op if `.is_finalized` is true, but then in this excerpt, `BuildResult` will immediately call `.finalize_build()`, which calls `self.atomic_dir.finalize()` (which would then be a no-op).

One possible way to refactor this might be to convert the `.finalized` check into an `@contextmanager`, or perhaps to provide some hook method when constructing an `AtomicDirectory` which is run when the finalization occurs.

I'm not super opinionated about this, and I think memorializing the kind of guarantee we want in the `finalized()` method and/or the `.is_finalized` property like @Eric-Arellano suggested would be sufficient here.

Reply:
> One possible way to refactor this might be to convert the `.finalized` check into an `@contextmanager`

That's done in https://github.com/jsirois/pex/pull/1 which I'll be making into a PR here as soon as I land this PR - which is based upon.

Reply:
Now up for review in #821 and the @contextmanager is here: https://github.com/pantsbuild/pex/pull/821/files#diff-4895aa16756686554f1906f607a1296eR272